src/components/Dialog.js

```diff
@@ -0,0 +1,139 @@
+import classnames from 'classnames';
+import { useEffect, useLayoutEffect, useRef, useState } from 'preact/hooks';
+
+import { IconButton, LabeledButton } from '../';
+import { SvgIcon } from './SvgIcon';
+
+let idCounter = 0;
+
+/**
+ * Return an element ID beginning with `prefix` that is unique per component instance.
+ *
+ * This avoids different instances of a component re-using the same ID.
+ *
+ * @param {string} prefix
+ */
+function useUniqueId(prefix) {
+  const [id] = useState(() => {
+    ++idCounter;
+    return `${prefix}-${idCounter}`;
+  });
+  return id;
+}
+
+/**
+ * @typedef {import("preact").ComponentChildren} Children
+ *
+ * @typedef DialogProps
+ * @prop {Children} [buttons] -
+ *   Additional `Button` elements to display at the bottom of the dialog.
+ *   A "Cancel" button is added automatically if the `onCancel` prop is set.
+ * @prop {string} [cancelLabel] - Label for the cancel button
+ * @prop {import("preact").ComponentChildren} children
+ * @prop {string} [contentClass] - CSS class to apply to the dialog's content
+ * @prop {string} [icon] - Name of optional icon to render in header
+ * @prop {import("preact/hooks").Ref<HTMLElement>} [initialFocus] -
+ *   Child element to focus when the dialog is rendered.
+ * @prop {() => void} [onCancel] -
+ *   A callback to invoke when the user cancels the dialog. If provided, a
+ *   "Cancel" button will be displayed.
+ * @prop {'dialog'|'alertdialog'} [role] - The aria role for the dialog (defaults to" dialog")
+ * @prop {string} title
+ */
+
+/**
+ * HTML control that can be disabled.
+ *
+ * @typedef {HTMLElement & { disabled: boolean }} InputElement
+ */
+
+/**
+ * Render a "panel"-like interface with a title and optional icon and/or
+ * close button. Grabs focus on initial render, defaulting to the entire
+ * Dialog container element, or `initialFocus` HTMLElement if provided.
+ *
+ * @param {DialogProps} props
+ */
+export function Dialog({
+  buttons,
+  cancelLabel = 'Cancel',
+  children,
+  contentClass,
+  icon,
+  initialFocus,
+  onCancel,
+  role = 'dialog',
+  title,
+}) {
+  const dialogDescriptionId = useUniqueId('dialog-description');
+  const dialogTitleId = useUniqueId('dialog-title');
+
+  const rootEl = useRef(/** @type {HTMLDivElement | null} */ (null));
+
+  useEffect(() => {
+    const focusEl = /** @type {InputElement|undefined} */ (initialFocus?.current);
```

Comment:
I thought `initialFocus` type was HTMLElement.

Reply:
I need some assistance! I inherited this code (it's from the `client`). I can see that this "cast" is intended to avoid a TS error on the line following:

`if (focusEl && !focusEl.disabled) {`

Only `InputElement`s have `disabled` props; `HTMLElement`s do not. I tried a few things but haven't been successful at finding a nice workaround. The intent here is: focus on whatever element was provided, but not if that element is disabled.  The provided ref/element might be an `InputElement`, but might not. I don't know the best way to solve this!